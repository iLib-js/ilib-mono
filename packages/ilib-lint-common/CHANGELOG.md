# ilib-lint-common

## 3.2.0

### Minor Changes

- 4a41879: - Updated Serializer API and clarified docs
  - Added the ability for SourceFile.write() to make the
    directories along the path where it wants to write
    the file if they don't exist already.
  - Added unit tests for SourceFile.write()
  - Updated Transformer API to add a results parameter
    to the transform() function so that the transformer
    can do things based on these results

## 3.1.2

### Patch Changes

- f9f1095: Unified package entrypoint definitions. This should help resolve edge cases where older packages (like Jest 26) were unable to correctly load some of them.
- ae2549a: Fixed broken links in package metadata.

## 3.1.1

### Patch Changes

- ff316a5: Linked dependencies within the monorepo.

## 3.1.0

- added formatOutput() method to the Formatter class
- added support for updating source files
  - made sure that SourceFile and IntermediateRepresentation instances are immutable
  - added the Transformer class to allow transforming the IntermediateRepresentation into a
    new IntermediateRepresentation. This allows multiple transformations to
    be applied to the file in sequence, such as filtering out any unwanted content,
    or adding a comment at the beginning of the file that tells the user that
    this is a generated file.
  - added the Serializer class to serialize the IntermediateRepresentation back into a SourceFile
  - added Plugin.getTransformers() method to allow plugins to define transformers
    that can be applied to the IntermediateRepresentation after the Rules are run
    on it
  - added Plugin.getSerializers() method to allow plugins to define serializers
    that can be applied to the IntermediateRepresentation after the Transformers
    are run on it
  - added PipelineElement class to ensure that Parser, Transformer, and Serializer
    classes all have the same interface in terms of names, descriptions, and
    types.

## 3.0.0

- Update major version because of these breaking changes which need to be
  updated in the plugins:
  - Added a SourceFile class to represent the original source file
    that is being parsed. This gives access to the raw bytes of the file,
    the file represented as a single string in Unicode, or to the file
    as an array of lines in Unicode.
  - Parser constructor no longer takes a filePath argument in the
    options object. Instead, the parse() method now takes a SourceFile
    argument. This means that ilib-lint will only create 1 instance of
    the parser instead of one per source file as it did before. It will
    then call the parse method multiple times with different source
    files to produce the intermediate representations of that file
  - The IntermediateRepresentation constructor no longer takes a filePath
    argument in the options object as well. Instead, it takes a SourceFile
    instance, giving the IntermediateRepresentation and any class that
    uses it access to the original raw content of the file or the
    representation of the file as a string or an array of lines. This will
    make it easier for Fixer and Fix instances to operate on the original
    file and for the Rules to generate the correct snippets of the original
    file to show where the issue occurred.
- renamed from i18nlint-common to ilib-lint-common to go along with the name of
  the tool

## 2.2.1

- changed the documentation for the return value of getRules and getFormatters
  so that it can be an array of classes or objects that definition declarative
  rules or formatters. This was already the case, but not documented properly.
- updated dependencies
- converted all unit tests from nodeunit to jest

## 2.2.0

- added Fix and Fixer to allow implementing auto-fixing of Results

## 2.1.0

- added the FileStats class
- added fields to the Result object for the location of the issue

## 2.0.1

- fix a problem where you could not create a Result instance with a
  lineNumber equal to 0.

## 2.0.0

- added Parser.getType() method
- clarified some jsdocs
  - Rule.match() should return a Result instance, an array of
    Result instances, or undefined if no match
- added IntermediateRepresentation class to represent the results of
  parsing a file
- Parser.parse() should now return the intermediate representation
  - this requires the plugins to change, so the major version is bumped
- Added support for logging provided by the lint tool so that plugins
  can use the linter's logging
  - getLogger() function passed to the constructors

## 1.4.0

- added utility function withVisibleWhitespace() to visually represent whitespace characters
- added isKebabCase(), isSnakeCase(), and isCamelCase() utility functions

## 1.3.0

- added getLink() method to the Rule class

## 1.2.0

- update the plugin to return only classes, as the linter may need to instantiate
  the classes multiple times
- add the getRuleSets() method to the plugin to allow the plugins to define
  standard rule sets

## 1.1.0

- added methods to abstract classes needed for loading and testing the plugins

## 1.0.0

- initial version
- define initial code and default built-in rules
