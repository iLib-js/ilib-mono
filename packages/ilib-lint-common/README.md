# ilib-lint-common

Common ilib-lint routines and classes/interfaces that the ilib-lint plugins
will need to implement the plugin functionality.


## Installation

```
npm install ilib-lint-common

or

yarn add ilib-lint-common
```

This package is usually imported by the ilib-lint tool itself or plugins for
the ilib-lint tool.

## Full API Docs

See the [full API docs](./docs/ilib-lint-common.md) for more information.

## License

Copyright Â© 2022-2024, JEDLSoft

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.

See the License for the specific language governing permissions and
limitations under the License.

## Release Notes

### v3.1.0
- added formatOutput() method to the Formatter class

### v3.0.0

- Update major version because of these breaking changes which need to be
  updated in the plugins:
    - Added a SourceFile class to represent the original source file
      that is being parsed. This gives access to the raw bytes of the file,
      the file represented as a single string in Unicode, or to the file
      as an array of lines in Unicode.
    - Parser constructor no longer takes a filePath argument in the
      options object. Instead, the parse() method now takes a SourceFile
      argument. This means that ilib-lint will only create 1 instance of
      the parser instead of one per source file as it did before. It will
      then call the parse method multiple times with different source
      files to produce the intermediate representations of that file
    - The IntermediateRepresentation constructor no longer takes a filePath
      argument in the options object as well. Instead, it takes a SourceFile
      instance, giving the IntermediateRepresentation and any class that
      uses it access to the original raw content of the file or the
      representation of the file as a string or an array of lines. This will
      make it easier for Fixer and Fix instances to operate on the original
      file and for the Rules to generate the correct snippets of the original
      file to show where the issue occurred.
- renamed from i18nlint-common to ilib-lint-common to go along with the name of
  the tool

### v2.2.1

- changed the documentation for the return value of getRules and getFormatters
  so that it can be an array of classes or objects that definition declarative
  rules or formatters. This was already the case, but not documented properly.
- updated dependencies
- converted all unit tests from nodeunit to jest

### v2.2.0

- added Fix and Fixer to allow implementing auto-fixing of Results

### v2.1.0

- added the FileStats class
- added fields to the Result object for the location of the issue

### v2.0.1

- fix a problem where you could not create a Result instance with a
  lineNumber equal to 0.

### v2.0.0

- added Parser.getType() method
- clarified some jsdocs
    - Rule.match() should return a Result instance, an array of
      Result instances, or undefined if no match
- added IntermediateRepresentation class to represent the results of
  parsing a file
- Parser.parse() should now return the intermediate representation
    - this requires the plugins to change, so the major version is bumped
- Added support for logging provided by the lint tool so that plugins
  can use the linter's logging
    - getLogger() function passed to the constructors

### v1.4.0

- added utility function withVisibleWhitespace() to visually represent whitespace characters
- added isKebabCase(), isSnakeCase(), and isCamelCase() utility functions

### v1.3.0

- added getLink() method to the Rule class

### v1.2.0

- update the plugin to return only classes, as the linter may need to instantiate
  the classes multiple times
- add the getRuleSets() method to the plugin to allow the plugins to define
  standard rule sets

### v1.1.0

- added methods to abstract classes needed for loading and testing the plugins

### v1.0.0

- initial version
- define initial code and default built-in rules
