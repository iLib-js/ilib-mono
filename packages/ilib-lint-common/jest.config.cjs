const { jestEsmConfig: baseConfig } = require("ilib-internal");

const config = {
    ...baseConfig,
    displayName: {
        name: "ilib-lint-common",
        color: "magentaBright",
    },
};

module.exports = config;
