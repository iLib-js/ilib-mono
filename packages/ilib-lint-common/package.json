{
    "name": "ilib-lint-common",
    "version": "3.1.0",
    "module": "./src/index.js",
    "type": "module",
    "exports": {
        ".": {
            "import": "./src/index.js"
        }
    },
    "description": "Shared code for the i18n linter",
    "keywords": [
        "internationalization",
        "i18n",
        "localization",
        "l10n",
        "globalization",
        "g11n",
        "lint",
        "locale"
    ],
    "homepage": "https://github.com/iLib-js/ilib-lint-common",
    "bugs": "https://github.com/iLib-js/ilib-lint-common/issues",
    "email": "marketing@translationcircle.com",
    "license": "Apache-2.0",
    "author": {
        "name": "Edwin Hoogerbeets",
        "web": "http://www.translationcircle.com/",
        "email": "edwin@translationcircle.com"
    },
    "contributors": [
        {
            "name": "Edwin Hoogerbeets",
            "email": "ehoogerbeets@gmail.com"
        },
        {
            "name": "Goun Lee",
            "email": "goun.lee@lge.com"
        }
    ],
    "files": [
        "src",
        "README.md",
        "LICENSE"
    ],
    "repository": {
        "type": "git",
        "url": "https://github.com/iLib-js/ilib-lint-common.git"
    },
    "engines": {
        "node": ">=14.0.0"
    },
    "scripts": {
        "build": "npm run build:prod",
        "build:prod": "grunt babel --mode=prod",
        "build:dev": "grunt babel --mode=dev",
        "build:test": "webpack-cli --config webpack-test.config.js",
        "dist": "npm-run-all doc build:prod build:pkg; npm pack",
        "test": "LANG=en_US.UTF8 node --experimental-vm-modules ./node_modules/.bin/jest",
        "test:watch": "LANG=en_US.UTF8 node --experimental-vm-modules ./node_modules/.bin/jest --watch",
        "debug": "LANG=en_US.UTF8 node --experimental-vm-modules --inspect-brk node_modules/.bin/jest -i",
        "clean": "git clean -f -d src test",
        "doc": "mkdir -p docs ; jsdoc2md -c jsdoc.json --separators --source src/* -m table > docs/ilib-lint-common.md ; npm run doc:html",
        "doc:html": "jsdoc -c jsdoc.json",
        "types": "tsc -p ./jsconfig.json"
    },
    "devDependencies": {
        "@tsconfig/node14": "^14.1.0",
        "@types/node": "^20.11.5",
        "docdash": "^2.0.2",
        "jest": "^29.7.0",
        "jsdoc": "^4.0.2",
        "jsdoc-to-markdown": "^8.0.0",
        "npm-run-all": "^4.1.5",
        "typescript": "^5.3.3"
    }
}
