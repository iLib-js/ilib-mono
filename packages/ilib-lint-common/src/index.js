/*
 * index.js - export everything from all of the files
 *
 * Copyright Â© 2022-2024 JEDLSoft
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import FileStats from './FileStats.js';
import Formatter from './Formatter.js';
import Parser from './Parser.js';
import Plugin from './Plugin.js';
import Result from './Result.js';
import Rule from './Rule.js';
import IntermediateRepresentation from './IntermediateRepresentation.js';
import Fix from './Fix.js';
import Fixer from './Fixer.js';
import SourceFile from './SourceFile.js';

export * from './utils.js';

export {
    FileStats,
    Formatter,
    Parser,
    Plugin,
    Result,
    Rule,
    IntermediateRepresentation,
    Fix,
    Fixer,
    SourceFile
};
