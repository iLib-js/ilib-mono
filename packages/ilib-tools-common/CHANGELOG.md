# ilib-tools-common

## 1.15.0

### Minor Changes

- 2e65e98: - Add getLanguagePluralCategories function to return an
  array of plural category names for the given ISO 639
  language code.
  - returns the categories from CLDR 47.0.0
  - example: "pl" -> ["one", "few", "many", "other"]

## 1.14.0

### Minor Changes

- c5ee237: - Added Escaper class to the tools library
  - escaperFactory() generates the appropriate escaper instance
    for the given programming language/syntax
  - Escaper.unescape() takes a string as it would be encoded in
    a source file, and transforms it into how it would be
    represented in memory while the program is running
  - Escaper.escape() does the converse
  - original code was extracted from various loctool plugins
  - this library is intended to be used with any loctool or ilib-lint
    plugin so that we do escaping/unescaping the same way no
    matter which plugin we are using

## 1.13.0

### Minor Changes

- 4965315: Added functions convertPluralResToICU and convertICUToPluralRes to convert
  ICU-style plural strings encoded in ResourceString instances into ResourcePlural
  instances and vice-versa
- 32ac524: - added directory walk routines
  - returns a js object that models the file system under the given directory
  - allows for arrays of include and exclude micromatch expressions

## 1.12.2

### Patch Changes

- f9f1095: Unified package entrypoint definitions. This should help resolve edge cases where older packages (like Jest 26) were unable to correctly load some of them.
- ae2549a: Fixed broken links in package metadata.
- Updated dependencies [ae2549a]
  - ilib-xliff@1.2.2

## 1.12.1

### Patch Changes

- ff316a5: Linked dependencies within the monorepo.
- Updated dependencies [ff316a5]
  - ilib-xliff@1.2.1

## 1.12.0

- Added support for the dnt flag (do not translate) in Resources
  - also for the "translate" flag of translation units

## 1.11.0

- added location information to the TranslationUnit and Resource constructors
  plus a Location class to keep track of the location in the source file
  where the resources come from.

## 1.10.0

- now ships with commonjs code as well as modern ESM in the same package
- updated dependencies

## 1.9.1

- forgot to export the HTML data in the previous version from the main entry point

## 1.9.0

- added data about HTML in the Utils
  - nonBreakingTags - a hash of all HTML tags that do not break a string
  - selfClosingTags - a hash of all HTML tags that are commonly self-closing
  - ignoreTags - a hash of HTML tags where the content is ignored, such as <script>
  - localizableAttributes - a hash of all tags that contain attributes which
    have localizable values

## 1.8.1

- update dependencies
- fixed a bug where the ResourceXliff.getVersion() call was documented to
  return a string, but it came out as a floating point number instead. Made
  it return the string properly.
- converted all unit tests to jest

## 1.8.0

- added parsePath() utility function which takes a template and a path
  and returns an object that maps each template parameter to a part of
  that path
  - getLocaleFromPath() is now re-implemented to use this
    function to find the locale parts of a path

## 1.7.0

- added getLines() method to tell how many lines there are in the xml file
- added support for location information of the start of each resource
  in the original file where the resource instances were read from
  - supports line and character within the line

## 1.6.0

- Added isDirty() method to the Resource class so we can see whether or
  not the resource has been modified since it was first created
  - also added clearDirty() method

## 1.5.0

- Added getVariant method to the TranslationUnit class

## 1.4.0

- Added TranslationUnit and TranslationVariant classes
- added hashKey function to the utilities
- fixed missing import for makeDirs() utility function

## 1.3.0

- Added more utility functions:
  - isEmpty - return whether or not an object is empty
  - cleanString - removing differences that are inconsequential for translation such as leading whitespace
  - makeDirs - create directories on disk
  - containsActualText - test if there is text left over after HTML and entities are stripped
  - objectMap - visitor pattern for objects

## 1.2.0

- Added formatPath and getLocaleFromPath utility function

## 1.1.0

- Added ResourceXliff class (represents an xliff file as a list of Resource instances)
- Added TranslationSet class (sets of Resources)
- Introduced some backwards compatibility support so that this library
  can be used with loctool plugins.
  - added some deprecated methods and accept some deprecated
    constructor parameters

## 1.0.0

- Initial code copied from loctool 2.18.0:
  - Resource
  - ResourceString
  - ResourceArray
  - ResourcePlural
