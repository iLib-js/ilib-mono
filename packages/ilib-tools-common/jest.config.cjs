const baseConfig = require("ilib-internal/jest.config.js");

const config = {
    ...baseConfig,
    displayName: {
        name: "ilib-tools-common",
        color: "blackBright",
    },
};

module.exports = config;
