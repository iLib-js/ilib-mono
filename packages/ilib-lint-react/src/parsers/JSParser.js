/*
 * JSParser.js - Parser for plain Javascript files
 *
 * Copyright Â© 2023-2024 Box, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import fs from 'fs';
import BabelParser from "@babel/parser";

import { Parser, IntermediateRepresentation } from 'ilib-lint-common';

/**
 * @class Parser for Javascript files based on the Babel parser.
 * By default, this parser will parse .js files. If you put your
 * javascript in a file with a different file name extension, you
 * can use the name "JSParser" in your filetype parsers array to
 * use this parser.
 */
class JSParser extends Parser {
    /**
     * Construct a new plugin.
     * @constructor
     */
    constructor(options = {}) {
        super(options);
        this.path = options.filePath;

        this.extensions = [ "js" ];
        this.name = "JSParser";
        this.description = "A parser for JS files.";
    }

    /**
     * @private
     */ 
    parseString(string, sourceFile) {
        return new IntermediateRepresentation({
            type: "babel-ast",
            ir: BabelParser.parse(string, {
                sourceType: "unambiguous"
            }),
            sourceFile
        });
    }

    /**
     * Parse the current file into an intermediate representation.
     * @param {SourceFile} sourceFile the source file to parse
     * @returns {Array.<IntermediateRepresentation>} the AST representation
     * of the jsx file
     */
    parse(sourceFile) {
        return [this.parseString(sourceFile.getContent(), sourceFile)];
    }

    getExtensions() {
        return this.extensions;
    }
};

export default JSParser;
