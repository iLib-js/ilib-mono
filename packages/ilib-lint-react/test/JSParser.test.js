/*
 * JSParser.test.js - test the React JS parser
 *
 * Copyright Â© 2023-2024 Box, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { ResourceString } from 'ilib-tools-common';

import JSParser from '../src/parsers/JSParser.js';

import { Result, IntermediateRepresentation, SourceFile } from 'ilib-lint-common';

describe("testJSParser", () => {
    test("JSParserConstructorEmpty", () => {
        expect.assertions(1);

        const parser = new JSParser();
        expect(parser).toBeTruthy();
    });

    test("JSParserGetDescription", () => {
        expect.assertions(2);

        const parser = new JSParser();
        expect(parser).toBeTruthy();

        expect(parser.getDescription()).toBe("A parser for JS files.");
    });

    test("JSParserGetName", () => {
        expect.assertions(2);

        const parser = new JSParser();
        expect(parser).toBeTruthy();

        expect(parser.getName()).toBe("JSParser");
    });

    test("JSParserGetExtensions", () => {
        expect.assertions(2);

        const parser = new JSParser();
        expect(parser).toBeTruthy();

        expect(parser.getExtensions()).toStrictEqual([ "js" ]);
    });

    test("JSParserSimple", () => {
        expect.assertions(3);

        const parser = new JSParser();
        expect(parser).toBeTruthy();

        const sourceFile = new SourceFile("x/y", {});
        const actual = parser.parseString("import foo from '../src/index.js';", sourceFile);
        expect(actual).toBeTruthy();
        const actualSimplified = JSON.parse(JSON.stringify(actual));

        expect(actualSimplified).toMatchSnapshot();
    });

    test("JSParserMoreComplex", () => {
        expect.assertions(3);

        const parser = new JSParser();
        expect(parser).toBeTruthy();

        const sourceFile = new SourceFile("x/y", {});
        const actual = parser.parseString(
            `// comment
            import foo from '../src/index.js';

            const str = "String";
            `, sourceFile);
        expect(actual).toBeTruthy();
        const actualSimplified = JSON.parse(JSON.stringify(actual));

        expect(actualSimplified).toMatchSnapshot();
    });
});

