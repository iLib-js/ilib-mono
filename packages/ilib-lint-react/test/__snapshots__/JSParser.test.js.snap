// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testJSParser JSParserMoreComplex 1`] = `
{
  "ir": {
    "comments": [
      {
        "end": 10,
        "loc": {
          "end": {
            "column": 10,
            "index": 10,
            "line": 1,
          },
          "start": {
            "column": 0,
            "index": 0,
            "line": 1,
          },
        },
        "start": 0,
        "type": "CommentLine",
        "value": " comment",
      },
    ],
    "end": 105,
    "errors": [],
    "loc": {
      "end": {
        "column": 12,
        "index": 105,
        "line": 5,
      },
      "start": {
        "column": 0,
        "index": 0,
        "line": 1,
      },
    },
    "program": {
      "body": [
        {
          "end": 57,
          "leadingComments": [
            {
              "end": 10,
              "loc": {
                "end": {
                  "column": 10,
                  "index": 10,
                  "line": 1,
                },
                "start": {
                  "column": 0,
                  "index": 0,
                  "line": 1,
                },
              },
              "start": 0,
              "type": "CommentLine",
              "value": " comment",
            },
          ],
          "loc": {
            "end": {
              "column": 46,
              "index": 57,
              "line": 2,
            },
            "start": {
              "column": 12,
              "index": 23,
              "line": 2,
            },
          },
          "source": {
            "end": 56,
            "extra": {
              "raw": "'../src/index.js'",
              "rawValue": "../src/index.js",
            },
            "loc": {
              "end": {
                "column": 45,
                "index": 56,
                "line": 2,
              },
              "start": {
                "column": 28,
                "index": 39,
                "line": 2,
              },
            },
            "start": 39,
            "type": "StringLiteral",
            "value": "../src/index.js",
          },
          "specifiers": [
            {
              "end": 33,
              "loc": {
                "end": {
                  "column": 22,
                  "index": 33,
                  "line": 2,
                },
                "start": {
                  "column": 19,
                  "index": 30,
                  "line": 2,
                },
              },
              "local": {
                "end": 33,
                "loc": {
                  "end": {
                    "column": 22,
                    "index": 33,
                    "line": 2,
                  },
                  "identifierName": "foo",
                  "start": {
                    "column": 19,
                    "index": 30,
                    "line": 2,
                  },
                },
                "name": "foo",
                "start": 30,
                "type": "Identifier",
              },
              "start": 30,
              "type": "ImportDefaultSpecifier",
            },
          ],
          "start": 23,
          "type": "ImportDeclaration",
        },
        {
          "declarations": [
            {
              "end": 91,
              "id": {
                "end": 80,
                "loc": {
                  "end": {
                    "column": 21,
                    "index": 80,
                    "line": 4,
                  },
                  "identifierName": "str",
                  "start": {
                    "column": 18,
                    "index": 77,
                    "line": 4,
                  },
                },
                "name": "str",
                "start": 77,
                "type": "Identifier",
              },
              "init": {
                "end": 91,
                "extra": {
                  "raw": ""String"",
                  "rawValue": "String",
                },
                "loc": {
                  "end": {
                    "column": 32,
                    "index": 91,
                    "line": 4,
                  },
                  "start": {
                    "column": 24,
                    "index": 83,
                    "line": 4,
                  },
                },
                "start": 83,
                "type": "StringLiteral",
                "value": "String",
              },
              "loc": {
                "end": {
                  "column": 32,
                  "index": 91,
                  "line": 4,
                },
                "start": {
                  "column": 18,
                  "index": 77,
                  "line": 4,
                },
              },
              "start": 77,
              "type": "VariableDeclarator",
            },
          ],
          "end": 92,
          "kind": "const",
          "loc": {
            "end": {
              "column": 33,
              "index": 92,
              "line": 4,
            },
            "start": {
              "column": 12,
              "index": 71,
              "line": 4,
            },
          },
          "start": 71,
          "type": "VariableDeclaration",
        },
      ],
      "directives": [],
      "end": 105,
      "interpreter": null,
      "loc": {
        "end": {
          "column": 12,
          "index": 105,
          "line": 5,
        },
        "start": {
          "column": 0,
          "index": 0,
          "line": 1,
        },
      },
      "sourceType": "module",
      "start": 0,
      "type": "Program",
    },
    "start": 0,
    "type": "File",
  },
  "sourceFile": {
    "dirty": false,
    "filePath": "x/y",
    "sourceLocale": "en-US",
    "type": "",
  },
  "type": "babel-ast",
}
`;

exports[`testJSParser JSParserSimple 1`] = `
{
  "ir": {
    "comments": [],
    "end": 34,
    "errors": [],
    "loc": {
      "end": {
        "column": 34,
        "index": 34,
        "line": 1,
      },
      "start": {
        "column": 0,
        "index": 0,
        "line": 1,
      },
    },
    "program": {
      "body": [
        {
          "end": 34,
          "loc": {
            "end": {
              "column": 34,
              "index": 34,
              "line": 1,
            },
            "start": {
              "column": 0,
              "index": 0,
              "line": 1,
            },
          },
          "source": {
            "end": 33,
            "extra": {
              "raw": "'../src/index.js'",
              "rawValue": "../src/index.js",
            },
            "loc": {
              "end": {
                "column": 33,
                "index": 33,
                "line": 1,
              },
              "start": {
                "column": 16,
                "index": 16,
                "line": 1,
              },
            },
            "start": 16,
            "type": "StringLiteral",
            "value": "../src/index.js",
          },
          "specifiers": [
            {
              "end": 10,
              "loc": {
                "end": {
                  "column": 10,
                  "index": 10,
                  "line": 1,
                },
                "start": {
                  "column": 7,
                  "index": 7,
                  "line": 1,
                },
              },
              "local": {
                "end": 10,
                "loc": {
                  "end": {
                    "column": 10,
                    "index": 10,
                    "line": 1,
                  },
                  "identifierName": "foo",
                  "start": {
                    "column": 7,
                    "index": 7,
                    "line": 1,
                  },
                },
                "name": "foo",
                "start": 7,
                "type": "Identifier",
              },
              "start": 7,
              "type": "ImportDefaultSpecifier",
            },
          ],
          "start": 0,
          "type": "ImportDeclaration",
        },
      ],
      "directives": [],
      "end": 34,
      "interpreter": null,
      "loc": {
        "end": {
          "column": 34,
          "index": 34,
          "line": 1,
        },
        "start": {
          "column": 0,
          "index": 0,
          "line": 1,
        },
      },
      "sourceType": "module",
      "start": 0,
      "type": "Program",
    },
    "start": 0,
    "type": "File",
  },
  "sourceFile": {
    "dirty": false,
    "filePath": "x/y",
    "sourceLocale": "en-US",
    "type": "",
  },
  "type": "babel-ast",
}
`;
